### ISSUE TITLE
Extremely sluggish scrolling

### ISSUE THREAD

**Issue by dcposch** [feross/webtorrent-desktop#725]
**What version of WebTorrent Desktop?** d0733d33705c4cbf5dee26c43cda2e8febc80419 (after 0.8.1)

**What operating system and version?** Mac OS El Capitan

**What did you do?** Opened the detail view on this torrent ([Huckleberry Finn](https://archive.org/details/huckleberry_mfs_librivox)), which contains 701 files.

**What did you expect to happen?** A long list of files with smooth scrolling.

**What actually happened?** Very choppy, nearly unusable scrolling and increased CPU load.

The app just froze for ~1 second at a time and became mostly unresponsive to inputs.

Collapsing the detail view made the app fast again.

The torrent wasn't finished downloading, so torrent progress events were causing the whole file list to redraw with new % complete for each file. Updating 700 `<divs>` once per second shouldn't be a problem, though.

**It looks like HyperX has poor performance.**

Here's the result of profiling during the situation described above:

<img width="1074" alt="screen shot 2016-07-17 at 5 53 22 pm" src="https://cloud.githubusercontent.com/assets/169280/16904357/2ab58dd8-4c49-11e6-9ae7-bb2b1329f344.png">

The biggest time sinks are all inside hyperx. It's parsing our HTML templates not only at runtime, but every single time they're called --- in this case, that means it's parsing the hyperx template in `renderFileRow()` 700+ times per call to update(). The parser is written in Javascript and is not particularly efficient.

The original and most popular virtual DOM library, React, parses its templates at compile time (JSX). That seems more reasonable.

(React also has other advantages. For example, our virtualdom + hyperx setup replaces every event listener in the app on every call to update() by default. We've partially worked around that by memoizing the event handlers using `dispatcher(...)`. In general React is a lot more widely used and more actively developed.)

--

WebTorrent Desktop needs to feel snappy and reliable. I suggest we do one of two things:
- Remove hyperx and go back to using virtualdom directly. (See https://github.com/feross/webtorrent-desktop/commit/b5561f13b4181e8ad9144fd882af0c7b5df5aac9)
- Switch to React

Thoughts?

**Comment by feross** [comment]
I've never used React on a project before, so I have no idea if I'll like it. At a high level, I agree with your reasoning. I'm happy to switch to it if it solves the perf issues.

**Comment by dcposch** [comment]
fixed!

switching from hyperx to React helped, but the biggest cause was really stupid. turns out the *download button CSS animation* was making the entire list slow.

i saved the app UI to a static HTML file, using the devtools inspector. then i opened that in browser and observed the same slow scrolling and slow hover highlights with no javascript running at all.

some stats:
- the file, torrent-details.html, is only 97KB
- the expanded torrent only has 350 files
- the table was already optimized
  - using `table-layout: fixed`, so no autosizing columns
  - fixed pixel height for rows
  - fixed pixel width for all but one column, which takes up the remaining width

... and despite that, here it is running at a blazing 10FPS:

![screen shot 2016-07-22 at 5 42 49 pm](https://cloud.githubusercontent.com/assets/169280/17075994/3d01d6ca-505b-11e6-8bc1-9c17966cc7a0.png)

notice how chrome *isn't even doing anything* for most of the 100ms it takes to render each frame. the devtools Timeline view has a useful, detailed breakdown of where time is spent--in the main thread, rasterizer threads, GPU, etc. but you can see most of the frame time, all of those are idle.

turns out that purple bar across the top is the continuously animating download button, the green arrow that bounces up and down. CSS animation, zero javascript. this is confined to a small element and has nothing to do with the file details table, and yet it makes the whole page very slow.

tldr; the DOM is heavy, CSS is complicated, and chrome has a bug :)

**Comment by feross** [comment]
👏
